### PR TITLE
has_error function

### LINKED ISSUE #3719: has_error()
We should have a has_error() function that returns true for any value that has an error somewhere inside of it.  Since errors must appear in the type signature, this could cache results in a type map.

### PR DESCRIPTION
**Pull request by mattnibs**
Add has_error function which will recurse a zed value and return true if
the value or any of its leafs an error. has_error memoizes the types it
has seen so it won't have to recurse types more than once (unless the it
contains an error type in a union in which case it'll always have to
recurse).

Closes #3719

### CODE REVIEW

**Comment by mattnibs** [comment]
Also of note, has_error doesn't support type values, should it?

File: runtime/expr/function/types.go [hunk 1]
```diff
@@ -107,6 +108,78 @@ func (i *Is) Call(_ zed.Allocator, args []zed.Value) *zed.Value {
...
+	var hasErr bool
+	canCache := true
+	switch typ := typ.(type) {
+	case *zed.TypeRecord:
+		it := b.Iter()
+		for _, col := range typ.Columns {
+			e, c := h.hasError(col.Type, it.Next())
+			hasErr = hasErr || e
+			canCache = !canCache || c
+		}
```
Comment (nwt) [inline runtime/expr/function/types.go:146]: This will panic on a null record value.

File: runtime/expr/function/types.go [hunk 2]
```diff
@@ -107,6 +108,78 @@ func (i *Is) Call(_ zed.Allocator, args []zed.Value) *zed.Value {
...
+
+func (h *HasError) Call(_ zed.Allocator, args []zed.Value) *zed.Value {
+	v := args[0]
+	if yes, _ := h.hasError(v.Type, v.Bytes); yes {
+		return zed.True
+	}
+	return zed.False
+}
+
+func (h *HasError) hasError(t zed.Type, b zcode.Bytes) (bool, bool) {
```
Comment (nwt) [inline runtime/expr/function/types.go:129]: I think you could simplify this if you used `Value.Walk`.

File: runtime/expr/function/types.go (hunk 1 above)
Comment (nwt) [reply runtime/expr/function/types.go:146]: This still panics on `null({a:int64})`.

Comment (nwt) [reply runtime/expr/function/types.go:146]: This still panics on `null({a:int64})`.

File: runtime/expr/function/types.go (hunk 2 above)
Comment (mattnibs) [reply runtime/expr/function/types.go:129]: I had considered this but there was some case that I thought had made it unworkable. Let me look again.

Comment (mattnibs) [reply runtime/expr/function/types.go:129]: Using walk is tougher because I need to keep track of whether a part of value tree is cacheable or not.

**Review by nwt — APPROVED** [review]

**Pull request merged**